fedbiomed/node/requests/_overlay.py

```diff
@@ -1,258 +1,542 @@
+    ready_event: asyncio.Event
+    distant_key: Optional[DHKey] = None
+
+
+class _ChannelKeys:
+    """Manages encryption/signing keys for each node to node channel
+
+    Attributes:
+        _channel_keys: (Dict[str, _N2nKeysEntry]) key status for node to node channels
+        _channel_keys_lock: (asyncio.Lock) lock to ensure exclusive access to _channel_keys
+    """
+    def __init__(self):
+        """Class constructor"""
+        self._channel_keys = {}
+        self._channel_keys_lock = asyncio.Lock()
+        self._channel_manager = ChannelManager(environ['DB_PATH'])
+
+        for distant_node_id in self._channel_manager.list():
+            channel = self._channel_manager.get(distant_node_id)
+
+            # don't need to acquire _channel_keys_lock in the constructor
+            self._channel_keys[distant_node_id] = _N2nKeysEntry(
+                local_key=DHKey(private_key_pem=channel['local_key']),
+                ready_event=asyncio.Event()
             )
-    except (FileNotFoundError, PermissionError, ValueError) as e:
-        raise FedbiomedNodeToNodeError(
-            f"{ErrorNumbers.FB324.value}: cannot read default node to node key "
-            f"from file {default_key_file}"
-        ) from e
-    return private_key
 
+    def _create_channel(self, distant_node_id: str) -> None:
+        """Create channel for peering with a given distant peer node if it does not exist yet.
 
-_default_n2n_key = load_default_n2n_key()
+        Assumes it is called with already acquired _channel_keys_lock
 
-# chunk size must be less or equal (in bits) the smallest RSA key length used
-_SMALLEST_KEY = 2048
-_CHUNK_SIZE = int(_SMALLEST_KEY / 8)
+        If channel does not exist, create it and generate local key.
+        If channel already exists, do nothing.
 
+        Args:
+            distant_node_id: unique ID of the peer node
+        """
+        if distant_node_id not in self._channel_keys:
+            local_key = DHKey()
+            self._channel_manager.add(distant_node_id, local_key.export_private_key())
 
-def format_outgoing_overlay(message: Message) -> List[bytes]:
-    """Creates an overlay message payload from an inner message.
+            self._channel_keys[distant_node_id] = _N2nKeysEntry(local_key=local_key, ready_event=asyncio.Event())
 
-    Serialize, crypt, sign the inner message
+    async def get_keys(self, distant_node_id: str) -> Tuple[DHKey, DHKey]:
+        """Gets keys for peering with a given distant peer node on a channel
 
-    Args:
-        message: Inner message to send as overlay payload
+        Args:
+            distant_node_id: unique ID of the peer node
 
-    Returns:
-        Payload for overlay message
+        Returns:
+            A tuple consisting of the public key object for the channel and the private key
+                object for the channel (or None if not known yet)
+        """
+        async with self._channel_keys_lock:
+            self._create_channel(distant_node_id)
```

Comment:
I think get keys should return `None` if distant node id is not existing ratheer than creating a channel. 

As I understand,  ChannelKeys have 3 responsiblities `get`, `add` and `synchronize (wait channel is ready)`. 
I would expect;  

1 - Instantiation: read database and store keys in memory
2 - Upper layer connect to node: 
       - Check if key already registered 
       - Yes: Return the key, otherwise `None`
       - In the upper layer if get key is None: Setup the key and create the channel, so `get_keys` does not create a channel  only `set_distant_key` creates the channel. 

Sorry but every method in this class creates channel, and it is quite difficult to follow the order or understand the scenario. 

**Another point:** 
Instead of returning `self._channel_keys[distant_node_id].local_key, self._channel_keys[distant_node_id].distant_key` you can return directly `self._channel_keys[distant_node_id]` that already allows to access local and distant key.  
`

Reply:
Thank you @srcansiz 

As discussed yesterday:
- I remove the `is_channel_ready` method, we will use `get_keys` instead
- returning `None` would bring additional burden to upper layer which needs to distinguish the case where the entry does not exist and the case where the entry exists but does not yet have a distant key

You are right that `_create_channel` would only be mandatory in:
- `is_ready_channel` (now replaced with `get_keys`) for calling from `_setup_use_channel_keys`. Or replace it by a `get_keys` + `create_channel` as you suggest
- `get_local_public_key` for answering to a ChannelSetupRequest message in n2n controller
- `set_distant_key` for handling a ChannelSetupReply message (which may update a non existing entry, if the distant node is confused or malicious)

In `wait_ready_channel` this is only for having a self-coherent class (for now it would only protect against un coherent code in `OverlayChannel`).